### DISCUSSION TITLE
Framework e/ou Lib JavaScript que estão utilizando/estudando?

### DISCUSSION THREAD

**Discussion by neysimoes** [frontendbr/forum#44] [Perguntas e Respostas]
Bom dia!
Estou começando um projeto pessoal e estou em dúvida do que usar para desenvolver. Queria saber o que vocês estão utilizando, as vantagens e desvantagens. Também se estão estudando alguma que ainda não usam.

Estava em dúvida entre Angular2 e ReactJS. :+1: 

Atualmente estou trabalhando num sistema interno da empresa feito em AngularJS, mas quando terminarmos vamos começar a refazer o e-commerce em ReactJS.

**Comment by bernardodiasc** [comment]
To trabalhando com React+Redux num projeto open source (ainda não posso mostrar, semana que vem sai a v1.0 ai divulgo). React é divertido, isso é importante. Fiz alguma coisa com Angular 1 (não conheço o 2 ainda), mas não curti muito, infelizmente não sei explicar exatamente o porquê.

React tem algumas coisinhas a se considerar: 
- faz mais sentido com stacks completas (react + flux + immutable + graphql + relay + blablabla)
- muito barulho e muita coisa nova todo dia, nem tudo é necessário ou importante

**Comment by willycamargo** [comment]
Se está em dúvida entre esses dois, recomendo [essa leitura](https://medium.freecodecamp.com/angular-2-versus-react-there-will-be-blood-66595faafd51#.yij0ivlul)

---

Spoiler:

```
Ember: {{# each}}
Angular 1: ng-repeat
Angular 2: ngFor
Knockout: data-bind=”foreach”
React: JUST USE JS. :)
```

**Comment by neysimoes** [comment]
@willycamargo foi exatamente esse texto que iniciou essa discussão :smiley:

**Comment by fdaciuk** [comment]
@neysimoes, se for usar o React, considere o [`preact`](https://github.com/developit/preact) e [`preact-compat`](https://github.com/developit/preact-compat) no lugar ;)

O `preact` tem a mesma interface do React, mas apenas 3kb! Nos testes que estou fazendo, ele está se saindo muito bem. E com o `preact-compat` + `webpack`, fica fácil manter o mesmo código já criado com `react` e `react-dom`, e mudar para o `preact`.

Só pra você ter uma ideia do tamanho dos arquivos finais, nas [brincadeiras que estou fazendo](https://github.com/fdaciuk/preact-playground), com React + Redux e Preact + Redux:

**Sem minificar:**
React = 712kb
Preact = 105kb

**Minificado:**
React = 209kb
Preact = 37.9kb

**Gzipped:**
React = 42.56kb
Preact = 10.49kb

:D

**Comment by simaodeveloper** [comment]
Atualmente estou estudando [VueJS](http://vuejs.org/) , ele é o cara que controla a View, mas se utilizarmos com algumas outras ferramentas, funciona como um ótimo framework.

Aqui vai uma [introdução ao VueJS](http://blog.evanyou.me/2015/10/25/vuejs-re-introduction/) e um repositório de links de coisas váriadas relacionadas ao Vue, se chama [Awesome VueJS](https://github.com/vuejs/awesome-vue)

Coverage 100.00% [veja o teste aqui](https://codecov.io/github/vuejs/vue?branch=master)

Exemplo: [Hacker News feito em Vue](https://github.com/vuejs/vue-hackernews)

A Comunidade construiu um router pra ele também [Vue-Router](http://vuejs.github.io/vue-router/en/index.html)

**Comment by suissa** [comment]
caraca hein @fdaciuk

**Comment by fdaciuk** [comment]
Loucura @suissa :shipit:

**Comment by eduardojmatos** [comment]
@fdaciuk ainda com a aba aberta aqui, do Preact. Vou pega pra brincar com isso essa semana :smiley: 

@neysimoes estou usando Ember no trabalho. Ainda me adaptando com as mágicas dele, mas até que curti :smile: . Recentemente tiraram o GlimmerComponent da versão Canary (do master do git), pra corrigir alguns bugs, mas é uma outra ideia interessante pra trabalhar com DOM diff.

Tô engatinhando nele, mas já curti bastante :smile:

**Comment by kazzkiq** [comment]
Uma outra alternativa _não-convencional_ ao React seria o [Riot.js](http://riotjs.com/). Estou usando ele em um projeto grandinho e até o momento to bem satisfeito tanto em termos de performance quanto com as features da biblioteca.

E tem a vantagem de ter uma sintaxe bem limpa e elegante IMO:

``` html
<!-- Componente contador numérico -->
<contador>
    <button onclick="{ sub }"> - </button>
    <input type="number" name="campo" readonly />
    <button onclick="{ add }"> + </button>

    <script>
        add() {
            this.campo.value = +this.campo.value + 1;
        }
        sub() {
            this.campo.value = +this.campo.value - 1;
        }
    </script>
</contador>
```

Para quem quiser dar uma brincada, [tem um playground aqui](http://riotjs.com/play/).

**Comment by ghost** [comment]
@kazzkiq que maneiro o Riot.js, não o conhecia, estou neste playground aí que passou.. é muito tranquilo, você participa dos forums? suporte da comunidade é bom? hehe

**Comment by kazzkiq** [comment]
@meche Participo, eles tem uma [sala no Gitter](https://gitter.im/riot/riot) e sempre tem gente tirando dúvidas lá. Versões novas com bugfixes/features seguem um ritmo bem frequente também. Até o atual momento, nada a reclamar.

**Comment by fabricionaweb** [comment]
Massa ver a galera de olho no **Riot**, também tenho brincado com ele! Super didático e está bombando no GitHub hehe

**Comment by felipefialho** [comment]
Que interessante esse Preact.

Já usou em projetos reais @fdaciuk?

**Comment by willycamargo** [comment]
Estamos com dois projetos SPA aqui. Um deles rodando com [React](https://facebook.github.io/react) e outro com o [MithrilJS](https://lhorie.github.io/mithril/index.html).

Ambas as libs (framework no caso do MithrilJS) são muito flexíveis.

Utilizo em ambas:
- [Fetch](https://github.com/github/fetch) para consumir os dados da API
- [Browserify](http://browserify.org/) com [Babel](https://babeljs.io/) para compilar o código ES6
- [Stylus](http://stylus-lang.com/) com [PostCSS](https://github.com/postcss/postcss) no CSS
- [Browsersync](https://www.browsersync.io/) para sincronizar os browsers

Utilizo o [Mocha](https://mochajs.org/) para testes unitários em um dos projetos e o [Jest](https://facebook.github.io/jest/) no outro.

---

No geral estamos satisfeitos com a produtividade e também com o desempenho do sistema nas duas situações.

**Comment by felisio** [comment]
Vou começar um projeto gigante com CRM modulo financeiro... Depois de muito pesquisar resolvi usar react + redux.... Quem já tem muita experiência nessa tecnologia acha tranquilo? Ou seria melhor adotar outra tecnologia? Menos angular kkkkkk

**Comment by fdaciuk** [comment]
@felisio acredito que esse seja o melhor caminho :)

**Comment by hugooliveirad** [comment]
Eu estou usando ClojureScript. Foge um pouco do título, mas acho legal citar.

Junto com ele uso basicamente duas libs: [reagent](https://reagent-project.github.io/) e [re-frame](https://github.com/Day8/re-frame).

Reagent é um wrapper para o React. Ele torna bem simples criar componentes como funções puras e otimiza bastante a renderização deles pois junta as estruturas de dados imutáveis do ClojureScript com o shouldComponentUpdate.

O re-frame tem uma ideia parecida com o [cerebral](https://github.com/cerebral/cerebral), um framework para fluxo de dados unidirecional. O legal é que ele usa o sistema de meta-programação do ClojureScript para criar uma relação de dependência entre seus dados e informações computadas. Bem como uma planilha deixa você criar funções dos valores de células.

Quem tem interesse em ver uma forma bem diferente de criar interfaces, sugiro dar uma olhada.

Com otimizações do ClojureScript, nosso App em produção na Xerpa esta com 173kb com gzip.

**Comment by tcelestino** [comment]
Comecei a estudar o [Meteor](https://www.meteor.com/), estou curtindo pra caramba. Sem contar que é fácil de implementar com outras bibliotecas JS.
